### PR TITLE
Introduce MDC aware ExecutorService

### PR DESCRIPTION
**Pull request by bhagyasakalanka**
### Proposed changes in this pull request

Current ExecutorService class is not MDC aware. Therefore, the logs will have wrong tenant details. This PR will introduce MDC aware ExecutorServices. With this, the executorService will be MDC aware and create new threads using the parent thread's MDC properties. 

In the future, it is recommended to use MDCAwareExecutors to get a new ExecutorService object instead of default Executors class

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/MDCAwareThreadPoolExecutor.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static org.wso2.carbon.identity.core.util.MDCUtils.wrapWithMdcContext;
+
+/**
+ * MDC aware thread pool executor. This will wrap override methods in an
+ * MDC aware manner to propagate correct values to newly created threads.
+ */
+
+public class MDCAwareThreadPoolExecutor extends ThreadPoolExecutor {
```
Comment (bhagyasakalanka) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/MDCAwareThreadPoolExecutor.java:35]: ThreadPoolAwareThreadPoolExecutor

**Review by darshanasbg — APPROVED** [review]

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ThreadLocalAwareExecutors.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Thread Local aware executors class. This will return a ExecutorService
+ * object which resolves MDC thread local in a multithreaded manner.
+ */
+public final class ThreadLocalAwareExecutors {
```
Comment (darshanasbg) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ThreadLocalAwareExecutors.java:30]: As we are introducing a new SPI, lets do a minor version bump.. Also, lets add unit tests..

**Review by darshanasbg — DISMISSED** [review]

**Comment by darshanasbg** [comment]
As we are introducing a new SPI, lets do a minor version bump.. Also, lets add unit tests..

**Comment by bhagyasakalanka** [comment]
> As we are introducing a new SPI, lets do a minor version bump.. Also, lets add unit tests..



> As we are introducing a new SPI, lets do a minor version bump.. Also, lets add unit tests..

Fixed

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/ThreadLocalAwareThreadPoolExecutorTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
+package org.wso2.carbon.identity.core;
```
Comment (darshanasbg) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/ThreadLocalAwareThreadPoolExecutorTest.java:1]: Add license headers

File: components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml [hunk 4]
```diff
@@ -27,6 +27,7 @@
             <class name="org.wso2.carbon.identity.core.dao.SAMLSSOServiceProviderDAOTest"/>
             <class name="org.wso2.carbon.identity.core.internal.DefaultServiceURLBuilderTest"/>
             <class name="org.wso2.carbon.identity.core.cache.BaseCacheTest"/>
+            <class name="org.wso2.carbon.identity.core.ThreadLocalAwareThreadPoolExecutorTest"></class>
```
Comment (darshanasbg) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml:30]: <class name="org.wso2.carbon.identity.core.ThreadLocalAwareThreadPoolExecutorTest"/>

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/ThreadLocalAwareThreadPoolExecutorTest.java (hunk 3 above)
Comment (bhagyasakalanka) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/ThreadLocalAwareThreadPoolExecutorTest.java:20]: fixed: 692c673ef089d55bb48c233dd8064eb69f989e7b

File: components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml (hunk 4 above)
Comment (bhagyasakalanka) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml]: fixed: 692c673ef089d55bb48c233dd8064eb69f989e7b

**Review by darshanasbg — APPROVED** [review]

**Review by ShanChathusanda93 — APPROVED** [review]

**Pull request merged**